tests/apps/social/test_pipeline_user.py

```diff
@@ -0,0 +1,89 @@
+
+from richie.apps.social.pipeline.user import USER_FIELDS, get_username
+
+
+class PipelineUserTestCase(TestCase):
+    """Tests for pipeline user."""
+
+    def test_get_username_no_username_in_backend(self):
+        """if username is not un backend settings should return None."""
+        backend = mock.Mock()
+        backend.setting.return_value = {}
+        strategy = mock.Mock()
+
+        self.assertIsNone(get_username(strategy, {}, backend))
+
+    def test_get_username_user_connected(self):
+        """The user is already connected, its username is fetch and returned."""
+        backend = mock.Mock()
+        backend.setting.return_value = USER_FIELDS
+        user = mock.Mock()
+
+        strategy = mock.Mock()
+        strategy.storage.user.get_username.return_value = "John Doe"
+
+        self.assertEqual(
+            get_username(strategy, {}, backend, user), {"username": "John Doe"}
+        )
+
+    def test_get_username_no_user_connected_email_as_username(self):
+        """No user connected and should return the email as username."""
+        backend = mock.Mock()
+        backend.setting.return_value = USER_FIELDS
+
+        strategy = mock.Mock()
+        strategy.setting.return_value = True
+        strategy.storage.user.user_exists.return_value = False
+
+        self.assertEqual(
+            get_username(strategy, {"email": "john@doe.org"}, backend),
+            {"username": "john@doe.org"},
+        )
+
+    def test_get_username_no_user_and_username_available(self):
+        """No user connected, the username is in the details and is available."""
+        backend = mock.Mock()
+        backend.setting.return_value = USER_FIELDS
+
+        strategy = mock.Mock()
+        strategy.setting.return_value = False
+        strategy.storage.user.user_exists.return_value = False
+
+        self.assertEqual(
+            get_username(strategy, {"username": "John Doe"}, backend),
+            {"username": "John Doe"},
+        )
+
+    def test_get_username_no_user_and_no_username_available(self):
+        """
+        No user connected and no username available in the details.
+        Should throw an exception.
+        """
+        backend = mock.Mock()
+        backend.setting.return_value = USER_FIELDS
+
+        strategy = mock.Mock()
+        strategy.setting.return_value = False
+
+        with self.assertRaises(AuthFailed):
+            get_username(strategy, {}, backend)
+
+    def test_get_username_no_user_and_username_already_saved(self):
+        """No user connected and the username is already saved. Should throw an exception."""
+        backend = mock.Mock()
+        backend.setting.return_value = USER_FIELDS
+
+        strategy = mock.Mock()
+        strategy.setting.return_value = False
+        strategy.storage.user.user_exists.return_value = True
+
+        with self.assertRaises(AuthAlreadyAssociated):
+            get_username(strategy, {"username": "John Doe"}, backend)
```

Comment:
I'm not really fan of these type of fully mocked unit tests. It helps for development but they will continue to pass when something changes in python social core... for example if the user_exists method disappears...
I feel creating a real user in db and testing the method with a real strategy and backend is much more security for this kind of override of an existing lib.

Reply:
Ok I will look how to do this.

Reply:
Done !